### PR TITLE
Add a progress bar for when scripts are running

### PR DESCRIPTION
**Pull request by dyc3**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Adds `r_print_progressbar_with_count`, and uses it to display progress when a script is being executed. The reason for not just changing `r_print_progressbar` is that I didn't want to break any existing functionality.

**Test plan**

```
r2 /bin/grep
Ps test
Po test
```

### CODE REVIEW

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6845]: This is already checked in while(x)

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6857]: I would add a config var to toggle this. I dont think we want progressbars all the time. Only for slow or large scripts maybe

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6884]: Uh?

Comment (trufae) [reply libr/core/cmd.c:6884]: Why not just flush after \r?

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd.c
Comment (dyc3) [reply libr/core/cmd.c:6884]: If this isn't present, the next line that is printed will go on the same line as the progress bar. Is there something else I should be using?

**Comment by dyc3** [comment]
I think a more ideal implementation would be to allow commands to submit progress so it could be rendered out separately with less overhead. It would allow us to clear the line when a new line is printed so we don't get extra progress bars, and allow multiple progress bars to be shown at once. However, I think this would require a significant rework of how printing works.

**Comment by dyc3** [comment]
I've been testing this out over the past couple of days, and one problem I'm having is that the progress bar shows up when piping r2's stdout somewhere. 2 potential solutions:
- Output the progress bar on stderr
- Disable the progress bar when output is piped

The other problem is that there's a series of 11 commands that gets run every time r2 starts up which causes the progress bar to appear, even when `scr.progressbar` is false. (Tested with `r2 -e scr.progressbar=false  /bin/ls` )

**Review by XVilka — CHANGES_REQUESTED** [review]
It broke the Windows tests

https://ci.appveyor.com/project/radareorg/radare2/builds/35802430/job/lys5v3rm6jfehvi6

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6841]: use size_t, and remove this unnecessary nesting braces

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6839]: use size_t too

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6842]: Just use  `537 R_API int r_str_char_count(const char *string, char ch) {`

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6851]: reduce the amount of duplicated code in here. something like this will look better:

```c
#if __UNIX__
 	 const bool istty = r_cons_isatty ();
#else
         const bool istty = true
 #endif

const bool show_progress_bar = core->print->enable_progressbar && r_config_get_i (core->config, "scr.interactive") && r_config_get_i (core->config, "scr.progressbar") && istty);
```

also move all this logic into a separate function so the condition can be checked in other places and the code will be more readable. at the end no #if code must be in here.

```
const show_progress_bar = want_progress_bar(core);
```

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6855]: sizet

File: libr/core/cmd.c
Comment (trufae) [inline libr/core/cmd.c:6890]: r_cons_newline(); or u dont want to use rcons buffering here?

File: libr/core/core.c
Comment (trufae) [inline libr/core/core.c:2734]: keep the old value and then set it back later, not force to true

File: libr/include/r_util/r_print.h
Comment (trufae) [inline libr/include/r_util/r_print.h:205]: R_API

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1591]: uhm there are two bugs in this division:

* total can be 0 and cause divbyzero exception
* pc is signed, so a division unrepresentable result can happen too

i wrote some macros few commits ago in master to detect such issues, use that before entering the loop. and add proper checks to avoid negative values in in cols/total/..

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1595]: use "%s", Color_RESET)

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1598]: first arg must be "%s"

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1592]: first arg must be "%s"

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1590]: maybe join this chk witt he if above. and avoid using the ternary operator for non assignments.
```c
if (enable_colors) {
  if (FLAGS_HEADER) {
    printf ("%s%4d%s%%...")
 }
  printf("[%s",Color_YELLOW)
} else {
  ...
}
```

File: libr/util/print.c
Comment (dyc3) [reply libr/util/print.c:1590]: I've made most of the requested changes, but I'm a little unsure about what to do for this. How exactly should I handle these cases? I'm not sure what macros you're referring to.

Comment (dyc3) [reply libr/util/print.c]: I've attempted to handle these cases adequately. The only macros I found were in `r_types_overflow.h`, and it wasn't clear to me how they would be used in this context.

**Comment by dyc3** [comment]
I don't know what I did to cause those windows builds to fail. They were working before...

File: libr/core/cconfig.c
Comment (trufae) [inline libr/core/cconfig.c:3709]: i think this option should be set to false by default it can be quite anoying for batch mode. or maybe make it honor the scr.interactive variable

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cconfig.c
Comment (dyc3) [reply libr/core/cconfig.c:3709]: I believe it does respect `scr.interactive`:

https://github.com/radareorg/radare2/blob/bd4c52f282ac3851b473a038508c55cf783af373/libr/core/cmd.c#L6852

What is the "batch mode" you're referring to?

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1626]: else goes here

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1611]: just oneline with ternary operator enable_colors?Color_YELLOW:""

also, instead of hardcoding the yellow and green. you may take the colors from the selected palette

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1602]: newline after `total,` make shorter lines

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1615]: cols * pc can result in negative value, making the loop iterate wrongly. You can solve this issue in different ways:

* use i > 0 instead of `i`
* compute cols*pc before the for body and check if the value is negative.
* use the OVFCHK apis

for this last thing you can use the following macros:

```
 65 UNSIGNED_DIV_OVERFLOW_CHECK(UT32_DIV_OVFCHK, ut32, UT32_MIN, UT32_MAX)
 66 UNSIGNED_DIV_OVERFLOW_CHECK(UT64_DIV_OVFCHK, ut64, UT64_MIN, UT64_MAX)
 71 SIGNED_MUL_OVERFLOW_CHECK(ST32_MUL_OVFCHK, st32, ST32_MIN, ST32_MAX)
 72 SIGNED_MUL_OVERFLOW_CHECK(ST64_MUL_OVFCHK, st64, ST64_MIN, ST64_MAX)
```

the problem here is that pc and cols are int and unsigned int. There are no macros for that, so probably the easiest way to go is checking if cols*pc > 0 before doing the for

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1621]: reuse the previous  cols*pc/total computation

File: libr/util/print.c
Comment (trufae) [inline libr/util/print.c:1585]: Use `hex.cols` eval var for this. use p->coreb.cfgGet(p->coreb->core, "hex.cols)` instead of hardcode the 78 in there. Bear in mind that hex.cols is 16 by default because its used for printing hexdumps. but as long as hexdumps are a common visual metric that variable is used to determine the width of other printed stuff.

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/util/print.c
Comment (dyc3) [reply libr/util/print.c:1611]: How do I grab the colors from the palette?

Comment (trufae) [reply libr/util/print.c:1611]: Use `r_cons_pal_get_i`, you can resolve the types by name with `r_cons_pal_get_name`. see cons/pal.c for all the possible values. The defines + enum in r_cons.h must be removed

**Comment by trufae** [comment]
Let's wait for the CI, lgtm :)

**Comment by dyc3** [comment]
I haven't made the changes you requested. Do you still want me to make those changes?

**Comment by trufae** [comment]
Just fix the ci

28453E0A-B711-403E-A853-F53EBBDE3B48

File: libr/util/print.c
Comment (trufae) [reply libr/util/print.c:1608]:
```suggestion
			p->cb_printf ("%s%4d%s%% %s%6d%s/%6d%s ", Color_GREEN, pc * 100 / total, Color_RESET, Color_GREEN, pc, Color_RESET, total, Color_YELLOW);
```

**Comment by trufae** [comment]
This branch cannot be rebased due to conflicts

**Comment by trufae** [comment]
As i expected its still far from perfect. but at least after 2months people is checking it and there are some discussions in the chat.

Screenshot 2020-12-29 at 15 59 09

**Pull request merged**
